_pytest/recwarn.py

```diff
@@ -13,83 +16,177 @@ def pytest_funcarg__recwarn(request):
-    return {'deprecated_call': deprecated_call}
+    return {'deprecated_call': deprecated_call,
+            'warns': warns}
 
 def deprecated_call(func, *args, **kwargs):
-    """ assert that calling ``func(*args, **kwargs)``
-    triggers a DeprecationWarning.
+    """Assert that ``func(*args, **kwargs)`` triggers a DeprecationWarning.
     """
-    l = []
-    oldwarn_explicit = getattr(warnings, 'warn_explicit')
-    def warn_explicit(*args, **kwargs):
-        l.append(args)
-        oldwarn_explicit(*args, **kwargs)
-    oldwarn = getattr(warnings, 'warn')
-    def warn(*args, **kwargs):
-        l.append(args)
-        oldwarn(*args, **kwargs)
-
-    warnings.warn_explicit = warn_explicit
-    warnings.warn = warn
+    wrec = WarningsRecorder()
+    wrec.start()
+    warnings.simplefilter('always')  # ensure all warnings are triggered
     try:
         ret = func(*args, **kwargs)
     finally:
-        warnings.warn_explicit = oldwarn_explicit
-        warnings.warn = oldwarn
-    if not l:
-        __tracebackhide__ = True
-        raise AssertionError("%r did not produce DeprecationWarning" %(func,))
+        wrec.finalize()
+
+    if not any(r.category is DeprecationWarning for r in wrec):
+        raise AssertionError("%r did not produce DeprecationWarning" % (func,))
+
     return ret
 
 
-class RecordedWarning:
-    def __init__(self, message, category, filename, lineno, line):
+def warns(expected_warning, *args, **kwargs):
+    """Assert that code raises a particular class of warning.
+
+    Specifically, the input @expected_warning can be a warning class or
+    tuple of warning classes, and the code must return that warning
+    (if a single class) or one of those warnings (if a tuple).
+
+    This helper produces a list of ``warnings.WarningMessage`` objects,
+    one for each warning raised.
+
+    This function can be used as a context manager, or any of the other ways
+    ``pytest.raises`` can be used::
+
+        >>> with warns(RuntimeWarning):
+        ...    warnings.warn("my warning", RuntimeWarning)
+    """
+
+    msg = ("exceptions must be old-style classes or"
+           " derived from Warning, not %s")
+    if isinstance(expected_warning, tuple):
+        for exc in expected_warning:
+            if not inspect.isclass(exc):
+                raise TypeError(msg % type(exc))
+    elif not inspect.isclass(expected_warning):
+        raise TypeError(msg % type(expected_warning))
+
+    wcheck = WarningsChecker(expected_warning)
+    if not args:
+        return wcheck
+    elif isinstance(args[0], str):
+        code, = args
+        assert isinstance(code, str)
+        frame = sys._getframe(1)
+        loc = frame.f_locals.copy()
+        loc.update(kwargs)
+
+        wcheck.start()
+        try:
+            code = py.code.Source(code).compile()
+            py.builtin.exec_(code, frame.f_globals, loc)
+        finally:
+            wcheck.finalize()
+    else:
+        func = args[0]
+        wcheck.start()
```

Comment:
Good point. I was still in the old mindset of trying to have this work without context managers, since that was in `pytest.raises`, but you're right that that's no longer necessary since we only support 2.6+.